Rotor output API

ahmed-elsaharti opened the pull request:
<!-- Thank you for submitting a pull request! -->
<!-- ⚠️⚠️ Do Not Delete This! pull_request_template ⚠️⚠️ -->
<!-- Please read our contribution guidelines: https://microsoft.github.io/AirSim/CONTRIBUTING/ -->

Potentially helps with : #672 #2419   <!-- add this line for each issue your PR solves. -->

## About
This PR adds an API that returns each rotor's speed, torque, and thrust. 
This is potentially useful for the future development of energy/battery level simulation within AirSim using @sytelus 's method outlined [here](https://github.com/microsoft/AirSim/issues/672#issuecomment-350904172)

~~The method used here is **super** crude and uses [a global variable](https://github.com/ahmed-elsaharti/AirSim/blob/03e044fe36eded643a54155e5fb0ed7b142bc8c8/AirLib/include/common/Common.hpp#L18-L24) that gets updated with the rotor parameters with every `updateRenderedState` while waiting for the `getRotorStates` API call. This probably is not the most elegant way of doing this.
**ie:**~~ Any testing and/or suggestions on how this could be improved are very welcome
Also, this is set up to work for quadrotors but scaling the struct up to take up any number of rotors is easy.


## How Has This Been Tested?
This was tested with the Blocks environment on UE 4.24 with Simpleflight and PX4 SITL. No issues so far.

## Screenshots:
On the python side here's the basic structure:
getRotorStates() (type = ~~RotorStateAPI~~ RotorStates) contains -> ~~rotors_ rotors (type = ~~RotorSTTVector~~ RotorVector) contains ->~~ rotors (type = list of rotors[0]-rotors[3] dicts) contains -> ['speed'], ['thrust'] and ['torque_scaler'] as floats

`print(client.getRotorStates())`:
![image](https://user-images.githubusercontent.com/49626509/102952892-cb815a00-4495-11eb-9efb-ab09981aff7d.png)

`print(client.getRotorStates().rotors_.rotor[0])`:
![image](https://user-images.githubusercontent.com/49626509/102952791-9412ad80-4495-11eb-8a49-771290cf3fc3.png)

ahmed-elsaharti committed:
parent adb2b8dc43d4c7ceb76cb7a1d5e180b8853c6673
author Ahmed Elsaharti <ahmed.elsaharti@und.edu> 1608383085 -0600
committer Ahmed Elsaharti <ahmed.elsaharti@und.edu> 1612289167 -0600

Added API

Added timestamp

added getRotorStates, cleaned up

Got rid of global var

Generalized RotorVector

zimmy87 commented on AirLib/include/api/RpcLibAdapatorsBase.hpp:
```diff
@@ -36,6 +36,62 @@ class RpcLibAdapatorsBase {
             d.push_back(TDest(s.at(i)));
     }
 
+    struct RotorSTTParameters {
```
Does STT here stand for "speed, thrust, torque_scalar"? If so, this strikes me running counter to the recommended ISO standard [here](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#nl5-avoid-encoding-type-information-in-names); i.e. I don't feel it's necessary to include information about a struct's members in its name.

ahmed-elsaharti replied:
Yes it does. Named it this way just for me to remember the purpose of the struct later. This was fixed in 3ee3cad anyway

zimmy87 commented on AirLib/include/api/RpcLibAdapatorsBase.hpp:
```diff
@@ -36,6 +36,62 @@ class RpcLibAdapatorsBase {
+            return d;
+        }
+    };
+    struct RotorSTTVector {
```
same comment as with RotorSTTParameters about possibly removing "STT" from the name

zimmy87 commented on AirLib/src/vehicles/multirotor/api/MultirotorApiBase.cpp:
```diff
@@ -11,6 +11,8 @@
 #include <iostream>
 #include <fstream>
 
+// Global var for rotor state API
+RotorParameters rotorparamsGlobal[4];
```
this can be removed if the recommended changes for replacing this global with a private variable are implemented elsewhere

zimmy87 reviewed (changes requested):
Hi @ahmed-elsaharti, thank you for creating this PR! I have a few comments that describe an alternative to using the global variable in your code. Most of my other comments deal with naming conventions and style guidelines. Please feel free to ask any questions if you'd like extra clarification about any of my comments.

ahmed-elsaharti commented on AirLib/include/common/CommonStructs.hpp:
```diff
@@ -8,6 +8,21 @@
 namespace msr { namespace airlib {
+//rotor feedback
+struct RotorSTTParameters {
+        real_T thrust = 0;
```
Fixed in 3ee3cad

ahmed-elsaharti commented on AirLib/include/common/CommonStructs.hpp:
```diff
@@ -8,6 +8,21 @@
+struct RotorSTTVector {
+    RotorSTTParameters rotor[4];
+
+        RotorSTTVector()
```
Fixed in 3ee3cad

ahmed-elsaharti commented on AirLib/include/vehicles/multirotor/api/MultirotorCommon.hpp:
```diff
@@ -16,7 +16,18 @@ enum class LandedState : uint {
 };
+// Struct for rotor state API
+struct RotorStateAPI {
+    RotorSTTVector rotors_;
```
Fixed in 3ee3cad

ahmed-elsaharti commented on AirLib/include/vehicles/multirotor/api/MultirotorRpcLibAdapators.hpp:
```diff
@@ -39,6 +39,27 @@ class MultirotorRpcLibAdapators : public RpcLibAdapatorsBase {
     };
 
+    struct RotorStateAPI {
+        RotorSTTVector rotors_;
```
Fixed in 3ee3cad

ahmed-elsaharti commented on PythonClient/airsim/types.py:
```diff
@@ -217,6 +217,8 @@ def nanPose():
         return (self.position.containsNan() or self.orientation.containsNan())
 
+class RotorSTTVector(MsgpackMixin):
+    Parameters = []
```
Fixed in 3ee3cad

ahmed-elsaharti commented on PythonClient/airsim/types.py:
```diff
@@ -359,6 +361,10 @@ class MultirotorState(MsgpackMixin):
 
+class RotorStateAPI(MsgpackMixin):
+    timestamp = np.uint64(0)
+    rotors_ = RotorSTTVector()
```
Fixed in 3ee3cad

ahmed-elsaharti commented on AirLib/include/vehicles/multirotor/api/MultirotorCommon.hpp:
```diff
@@ -16,7 +16,18 @@ enum class LandedState : uint {
     Flying = 1
 };
+// Struct for rotor state API
+struct RotorStateAPI {
```
Fixed in 3ee3cad

ahmed-elsaharti commented on AirLib/include/vehicles/multirotor/api/MultirotorRpcLibAdapators.hpp:
```diff
@@ -39,6 +39,27 @@ class MultirotorRpcLibAdapators : public RpcLibAdapatorsBase {
         }
     };
 
+    struct RotorStateAPI {
```
Fixed in 3ee3cad

ahmed-elsaharti commented on PythonClient/airsim/client.py:
```diff
@@ -1231,7 +1231,17 @@ def getMultirotorState(self, vehicle_name = ''):
+    def getRotorStates(self, vehicle_name = ''):
+        """
+        Args:
+            vehicle_name (str, optional): Vehicle to get the rotor state of
```
Added more elaboration on what the API is returning. Please let me know if more/less detail is required

ahmed-elsaharti commented on Unreal/Plugins/AirSim/Source/Vehicles/Multirotor/MultirotorPawnSimApi.cpp:
```diff
@@ -74,6 +74,10 @@ void MultirotorPawnSimApi::updateRenderedState(float dt)
+        // update global rotor variable
+        rotorparamsGlobal[i].speed = rotor_output.speed;
+        rotorparamsGlobal[i].thrust = rotor_output.thrust;
+        rotorparamsGlobal[i].torque_scaler = rotor_output.torque_scaler;
```
Thanks for the elaborate instructions David :D. I implemented the methodology you described in e67c2b2.

ahmed-elsaharti commented on AirLib/include/common/Common.hpp:
```diff
@@ -15,6 +15,14 @@
+    float thrust;
+};
+extern RotorParameters rotorparamsGlobal[4];
+
```
Global variable and struct have been removed 👍

ahmed-elsaharti commented on AirLib/include/api/RpcLibAdapatorsBase.hpp:
```diff
@@ -36,6 +36,62 @@ class RpcLibAdapatorsBase {
+        }
+    };
+    struct RotorSTTVector {
+        RotorSTTParameters rotor[4];
```
Thats a very valid point! 
To get this to work without memory access violations I had to clear the vector (might not be necessary but I did it as a precaution) and fill it with the rotor params in both [MultirotorPawnSimApi.cpp](https://github.com/microsoft/AirSim/blob/83e3a6dc1901deb5f10cb22e326316039dabd527/Unreal/Plugins/AirSim/Source/Vehicles/Multirotor/MultirotorPawnSimApi.cpp#L75-L79) and [RpcLibAdapatorsBase.hpp](https://github.com/microsoft/AirSim/blob/83e3a6dc1901deb5f10cb22e326316039dabd527/AirLib/include/api/RpcLibAdapatorsBase.hpp#L77-L80). I'm pretty sure there's a more elegant way of doing this though. Let me know if this needs changing

ahmed-elsaharti commented:
Hi @zimmy87, Thank YOU for the EXTREMELY helpful comments!
I fixed all the naming issues, got rid of the global variable and generalized the size of 'rotors'.
Please let me know if any other modifications are needed.

zimmy87 commented:
Hi @ahmed-elsaharti, 2 of the checks are failing for this PR due to a unrelated break in the Unity build from another PR. The fix for this build break was committed earlier today, can you rebase this PR to pick up the fix and rerun all the checks?

ahmed-elsaharti commented:
@zimmy87 Yup, that PR fixed the Unity build and the checks have passed 👍

zimmy87 commented on AirLib/include/api/RpcLibAdapatorsBase.hpp:
```diff
@@ -36,6 +36,64 @@ class RpcLibAdapatorsBase {
+        msr::airlib::RotorVector to() const
+        {
+            msr::airlib::RotorVector d;
+            for (unsigned int i = 0; i < d.rotor.size(); i++)
```
shouldn't this for loop look like this:
```
for (unsigned int i = 0; i < rotor.size(); i++)
```
I think as is, d.rotor.size() will always evaluate to 0. Also you need to push a new RotorParameters() on to d.rotor inside the for loop.

ahmed-elsaharti replied:
That actually makes complete sense.
Let me know if the latest commit does what you had in mind.

zimmy87 reviewed (changes requested):
This PR looks good for the most part, I just have 1 remaining comment.

ahmed-elsaharti committed:
Update RpcLibAdapatorsBase.hpp

rajat2004 commented on AirLib/include/vehicles/multirotor/api/MultirotorApiBase.hpp:
```diff
@@ -143,6 +148,12 @@ class MultirotorApiBase : public VehicleApiBase {
     }
 
+    /******************* rotors' states setter ********************/
+    void setRotorStates(RotorStates rotor_states)
```
Pass by ref here?

ahmed-elsaharti replied:
Fixed

rajat2004 commented on AirLib/include/api/RpcLibAdapatorsBase.hpp:
```diff
@@ -36,6 +36,65 @@ class RpcLibAdapatorsBase {
+        RotorVector(const msr::airlib::RotorVector& s)
+        {
+            rotor.clear();
+            for (unsigned int i = 0; i < s.rotor.size(); i++)
+            {
+                rotor.push_back(RotorParameters());
+                rotor[i] = s.rotor[i];
+            }
```
This can be replaced by something like (variable names can be improved ig) - 
```
for(const auto& r : s.rotor)
    rotor.push_back(RotorParameters(r));
```
Looks cleaner and avoids copy constructors

zimmy87 replied:
if RotorVector is removed, I still think this is a good change to make in msr::airlib_rpclib::MultirotorRpcLibAdapators::RotorStates

ahmed-elsaharti replied:
Removed RotorVector and applied these changes to msr::airlib_rpclib::MultirotorRpcLibAdapators::RotorStates

rajat2004 commented on AirLib/include/api/RpcLibAdapatorsBase.hpp:
```diff
@@ -36,6 +36,65 @@ class RpcLibAdapatorsBase {
+        msr::airlib::RotorVector to() const
+        {
+            msr::airlib::RotorVector d;
+            for (unsigned int i = 0; i < rotor.size(); i++)
+            {
+                d.rotor.push_back(msr::airlib::RotorParameters());
+                d.rotor[i].speed = rotor[i].speed;
+                d.rotor[i].thrust = rotor[i].thrust;
+                d.rotor[i].torque_scaler = rotor[i].torque_scaler;
+            }
```
Same, something like -
```
for(const auto& r : rotor)
    d.rotor.push_back(r.to());
```
Copying each value individually will just cause more pain if the struct needs to be modified later on

zimmy87 replied:
same as above, I think this will end up applying to msr::airlib_rpclib::MultirotorRpcLibAdapators::RotorStates

ahmed-elsaharti replied:
Removed RotorVector and applied these changes to msr::airlib_rpclib::MultirotorRpcLibAdapators::RotorStates

rajat2004 commented on Unreal/Plugins/AirSim/Source/Vehicles/Multirotor/MultirotorPawnSimApi.cpp:
```diff
@@ -68,8 +68,14 @@ void MultirotorPawnSimApi::updateRenderedState(float dt)
     for (unsigned int i = 0; i < rotor_count_; ++i) {
         const auto& rotor_output = multirotor_physics_body_->getRotorOutput(i);
+        // update private rotor variable
+        rotor_states_.rotors.rotor.push_back(RotorParameters());
+        rotor_states_.rotors.rotor[i].speed = rotor_output.speed;
+        rotor_states_.rotors.rotor[i].thrust = rotor_output.thrust;
+        rotor_states_.rotors.rotor[i].torque_scaler = rotor_output.torque_scaler;
```
Adding the parameterized constructor will allow cleaning up this part as well

ahmed-elsaharti replied:
Similar 'big picture' concept as [this](https://github.com/microsoft/AirSim/pull/3242#discussion_r570335117), although this takes up 4 lines and could probably be cleaner, I think keeping it simple makes this easy to look at and understand later for those (like me) who haven't dived deeply into AirSim's structure.

zimmy87 replied:
While making the setting of these values more explicit can make it easier to understand for someone who is just looking at this file, it also makes the code slightly harder to maintain. For example, if a new field was added to RotorParameters and we wanted to keep it updated along with all the other fields, we would also have to know to change this file to ensure the update happens and we wouldn't necessarily catch that until we notice a bug at runtime. Whereas when updating a parameterized constructor for a new field, leaving this file as is will cause a compile error, forcing the developer to keep this file updated to match the RotorParameters definition. Additionally, I think most contributors will either recognize what the parameterized constructor is doing, or they have access to an IDE with auto-lookup so it would be rather straight-forward to lookup the definition of the constructor if they are confused. In this case, I would prefer to take the slight hit to readability as a trade-off for extra maintainability.

ahmed-elsaharti replied:
Added an initializer list to RotorParameters and replaced these with [the if condition below](https://github.com/microsoft/AirSim/pull/3242#discussion_r580833347) and some description of what's happening

rajat2004 commented on AirLib/include/common/CommonStructs.hpp:
```diff
@@ -8,6 +8,21 @@
+    {}
+};
+struct RotorVector {
+    std::vector<RotorParameters> rotor;
```
Instead of just `rotor`, maybe something like `rotorv`? This line seemed pretty confusing in `MultirotorPawnSimApi` - `rotor_states_.rotors.rotor[i].speed`. TBH, I don't really see the point of having a separate `RotorVector` struct, but that's just my opinion and depends on the maintainers

ahmed-elsaharti replied:
Honestly although it looks odd, I did it with rotors.rotor[i] just to keep everything separate as is the case with most of AirSim's code. Just so that if the code needs to be changed or is looked at later by someone else they'd know that its a parent struct containing vectors of rotor parameters.
This also would help make things easier to understand/keep-track-of on the end user's side imo.

ahmed-elsaharti replied:
Got rid of RotorVector 👍

rajat2004 commented on PythonClient/airsim/types.py:
```diff
@@ -359,6 +361,10 @@ class MultirotorState(MsgpackMixin):
 
+class RotorStates(MsgpackMixin):
+    timestamp = np.uint64(0)
+    rotors = RotorVector()
```
This could be moved to near the `RotorVector` def

ahmed-elsaharti replied:
Mainly placed right below Multirotor state while RotorVector is placed at line 220 to stay consistent with (what I think is) the format of the types.py file. Where classes at the top contain classes like quaternion, etc, etc and the ones at the bottom contain classes with variables of the type of these top classes.

ahmed-elsaharti replied:
Got rid of RotorVector 👍

rajat2004 commented on Unreal/Plugins/AirSim/Source/Vehicles/Multirotor/MultirotorPawnSimApi.cpp:
```diff
@@ -68,8 +68,14 @@ void MultirotorPawnSimApi::updateRenderedState(float dt)
     collision_response = multirotor_physics_body_->getCollisionResponseInfo();
 
     //update rotor poses
+    rotor_states_.rotors.rotor.clear();
     for (unsigned int i = 0; i < rotor_count_; ++i) {
         const auto& rotor_output = multirotor_physics_body_->getRotorOutput(i);
+        // update private rotor variable
+        rotor_states_.rotors.rotor.push_back(RotorParameters());
```
Instead of continuously clearing and pushing back new objects, can't there instead be like a `set` method or something which would just update the values

ahmed-elsaharti replied:
Fixed

zimmy87 commented on AirLib/src/vehicles/multirotor/api/MultirotorApiBase.cpp:
```diff
@@ -849,6 +849,5 @@ float MultirotorApiBase::getObsAvoidanceVelocity(float risk_dist, float max_obs_
     unused(risk_dist);
     return max_obs_avoidance_vel;
 }
-
```
Is there any particular reason for removing the whitespace here? Seems like this might be left-over from a previous change.

ahmed-elsaharti replied:
Fixed

zimmy87 reviewed (changes requested):
I agree with most of @rajat2004's comments and think it would be prudent to include them before merging

ahmed-elsaharti committed:
Got rid of redundant struct

ahmed-elsaharti committed:
added an initializer list to RotorParameters

ahmed-elsaharti commented on AirLib/include/common/CommonStructs.hpp:
```diff
@@ -8,6 +8,21 @@
+    real_T speed = 0;
+
+    RotorParameters()
+    {}
```
Valid point. An initializer list has been added to RotorParameters

ahmed-elsaharti committed:
removed clear()

ahmed-elsaharti commented:
@rajat2004 @zimmy87 Thank you guys for the useful comments!
I can't for the life of me remember why I added RotorVector initially but I do agree, it is 100% redundant at this point.
I adjusted everything to:
![image](https://user-images.githubusercontent.com/49626509/109210995-3c354300-7773-11eb-8a8c-8dc185cccf21.png)
rather than:
![image](https://user-images.githubusercontent.com/49626509/109211110-62f37980-7773-11eb-93d6-fc94a7fce711.png)
and added an initializer list to `RotorParameters` to keep things clean when populating/updating the rotor states.
One thing I know for sure is that I should stop looking at PR reviews after midnight. Everything made much more sense when I re-read the comments a few days later 😅😅

Anyway, the PR should be good to go now that the Github checks have run successfully. Let me know if any other modifications are needed.

rajat2004 commented on Unreal/Plugins/AirSim/Source/Vehicles/Multirotor/MultirotorPawnSimApi.cpp:
```diff
@@ -70,6 +70,13 @@ void MultirotorPawnSimApi::updateRenderedState(float dt)
+            rotor_states_.rotors.push_back(RotorParameters(rotor_output.speed, rotor_output.thrust, rotor_output.torque_scaler));
+        else
+            //if the vector has already been populated previously, just update the values
+            rotor_states_.rotors[i] = RotorParameters(rotor_output.speed, rotor_output.thrust, rotor_output.torque_scaler);
```
This is still creating and replacing a new object, a more efficient way could be something like `rotor_states_.rotors[i].update(rotor_output.speed, rotor_output.thrust, rotor_output.torque_scaler)`, where `update` or `setValues` would just update the values.

Just my 2 cents, its up to the PR author and the maintainers as to which is better

ahmed-elsaharti replied:
I'm honestly fine with both ways. 
@zimmy87 what do you think? as-is or should I add an `update` method to `RotorParameters`.

I've also realized that I made a mistake here with the order of the parameters. That has been fixed in the latest commit

ahmed-elsaharti replied:
On the topic of efficiency, I was thinking of having a for loop populate `rotor_states_.rotors` in `MultirotorPawnSimApi::initialize` once instead of having to go through an if condition (the condition for-which would always be false after the first run) with every call of `MultirotorPawnSimApi::updateRenderedState`. Any thoughts?

rajat2004 replied:
Yeah, that should be a good way, similar to [this](https://github.com/microsoft/AirSim/blob/0a3f84c92e195f1d4b217afb5dcedfea09bff454/Unreal/Plugins/AirSim/Source/Vehicles/Multirotor/MultirotorPawnSimApi.cpp#L27) should work

ahmed-elsaharti replied:
Oh yeah, using the `assign` modifier is a better idea 👍

zimmy87 replied:
I like the idea of using the assign function in `MultirotorPawnSimApi::initialize` to instantiate all the structs and then calling an `update` method defined in `RotorParameters` to set the values in each struct here

ahmed-elsaharti replied:
Done 👍 
Let me know if anything else needs changing

rajat2004 commented on PythonClient/airsim/client.py:
```diff
@@ -1282,7 +1282,20 @@ def getMultirotorState(self, vehicle_name = ''):
+            vehicle_name (str, optional): Vehicle to get the rotor state of
 
+        Returns:
+            RotorStates: Containing a timestamp and the speed, trust and torque of all rotors.
```
Spelling - `thrust`

ahmed-elsaharti replied:
Good catch!
Fixed

rajat2004 reviewed (commented):
Just some small nitpicks

ahmed-elsaharti committed:
Minor changes, fixed rotor parameter order

ahmed-elsaharti commented on AirLib/include/vehicles/multirotor/api/MultirotorApiBase.hpp:
```diff
@@ -143,6 +148,12 @@ class MultirotorApiBase : public VehicleApiBase {
     }
 
+    /******************* rotors' states setter ********************/
+    void setRotorStates(RotorStates& rotor_states)
```
Fixed

ahmed-elsaharti commented on AirLib/include/vehicles/multirotor/api/MultirotorRpcLibAdapators.hpp:
```diff
@@ -39,6 +39,35 @@ class MultirotorRpcLibAdapators : public RpcLibAdapatorsBase {
+            std::vector<msr::airlib::RotorParameters> d;
+            for (const auto& r : rotors)
+            {
+                d.push_back(msr::airlib::RotorParameters(r.to()));
```
Fixed

ahmed-elsaharti commented on AirLib/include/common/CommonStructs.hpp:
```diff
@@ -8,6 +8,20 @@
 
 namespace msr { namespace airlib {
+//rotor feedback
+struct RotorParameters {
```
Good point. Fixed

ahmed-elsaharti commented on AirLib/include/api/RpcLibAdapatorsBase.hpp:
```diff
@@ -36,6 +36,29 @@ class RpcLibAdapatorsBase {
             d.push_back(TDest(s.at(i)));
     }
 
+    struct RotorParameters {
```
Fixed

zimmy87 reviewed (approved):
This PR looks good to me and once the one remaining feedback concerning populating the rotor states is resolved, I think it will be ready for merging

ahmed-elsaharti committed:
minor changes

ahmed-elsaharti commented:
@zimmy87 the last commit (4d8b3c9) should take care of [these changes](https://github.com/microsoft/AirSim/pull/3242#discussion_r585914035). Let me know if any other changes are needed 👍

zimmy87 commented:
Thank you for the contribution @ahmed-elsaharti!

condeLucas commented:
Hi @ahmed-elsaharti

Thank you for your contribution. 

Using @sytelus 's method to have the energy/battery consumption, we have only to multiply **Thrust** * **torque_scaler** * **speed** for each rotor and sum then during the simulation time?

Do you know the units of these measurements?

Thank you so much for your attention.
